Added GPO check and fixed other issues

ykuijs opened the pull request:
This PR implements a GPO check to prevent an endless reboot loop, as described in issue #13.

At the same time, I have also fixed issue #3 and #6.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xcredssp/14)
<!-- Reviewable:end -->

msftclas commented:
@ykuijs,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

ykuijs commented:
Added fix for issue #8

kwirkykat commented:
There aren't any end-to-end tests for this resource. Have you tested it manually?

---

Reviewed 2 of 3 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[README.md, line 45 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZWMAvh0Y5tGflqwg:-KeuZWMAvh0Y5tGflqwh:b-ysyrti) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L45)):*
> ```Markdown
> *   **Ensure:** Specifies whether the domain trust is present or absent 
> *   **Role**: REQUIRED parameter representing the CredSSP role, and is either "Server" or "Client" 
> *   **DelegateComputers**: Array of servers to be delegated to, REQUIRED when Role is set to "Client".
> ```

Can you please also update this section since there is a new parameter?

---

*[README.md, line 52 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZM58gi4I1rGu-1uo:-KeuZM58gi4I1rGu-1up:b5h6j5n) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L52)):*
> ```Markdown
> * Converted appveyor.yml to install Pester from PSGallery instead of from Chocolatey.
> * Implemented a GPO check to prevent an endless reboot loop when CredSSP is configured via a GPO
> * Fixed issue with Test always returning false with other regional settings then english
> ```

How do these changes fix this issue with Test-TargetResource?

---

*[README.md, line 53 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZ7akEL8Pvnseh-mG:-KeuZ7akEL8Pvnseh-mH:bovvuq) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L53)):*
> is Role=Server

if Role=Server

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 13 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuTJBu9vxJXJWG1xk3:-KeuTJBu9vxJXJWG1xk4:b-lw26f) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L13)):*
> $DelegateComputers

This variable/parameter doesn't exist here.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 146 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuWYD_sCYFXIg11CLW:-KeuWYD_sCYFXIg11CLX:b-7jgzww) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L146)):*
> ($DelegateComputers)

This should probably be $PSBoundParameters.ContainsKey('DelegateComputers')
Very minor. Only covers the extra case in which someone somehow managed to pass in $null or some mysterious value that translates to false for DelegateComputers

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 311 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuXJ7Uh8WAiNEc_4h0:-KeuXJ7VO2a4vwS1_nWi:b-a5ikfz) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L311)):*
> Get-ItemProperty -Path $Path -ErrorAction SilentlyContinue

You should probably check that this is not null first here so that it doesn't throw an error that you are trying to call a method/property on a null-valued expression

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.schema.mof, line 7 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuU7JezuMAjYA0_RYi:-KeuU7JezuMAjYA0_RYj:bcs9ta) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.schema.mof#L7)):*
> is

if

---

*[Tests/Unit/MSFT_xCredSSP.tests.ps1, line 107 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuXef_1-Q6z-s1vYxo:-KeuXef_1-Q6z-s1vYxp:b-ltx2ns) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/Tests/Unit/MSFT_xCredSSP.tests.ps1#L107)):*
> ```PowerShell
> 
>                 it 'should throw' {
>                     {Set-TargetResource -Ensure 'Present' -Role Server}| should throw 
> ```

I know this is how the previous tests here are written, but they are actually wrong as well. These calls to Assert-MockCalled aren't actually checking anything because the call to Set-TargetResource is inside braces ({}) which is considered a different scope. Weird Pester manipulation. Anyways, it's very easy to fix. Just add -Scope 'Context' to the calls to Assert-MockCalled. This will signal to Pester to check for the mock call inside any scope that was called within the Context rather than just the top-level scope of the Context. Like this:

`Assert-MockCalled -CommandName Enable-WSManCredSSP -Times 0 -Scope 'Context'`

The tests are passing at the moment because these are 'not' mock called checks, and since Pester thinks that the function was never even called within the current scope, the mocks were not called either.

---

*[Tests/Unit/MSFT_xCredSSP.tests.ps1, line 132 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-Keu_GjBWj59TofmHfg1:-Keu_GjBWj59TofmHfg2:b-rf3fcx) ([raw file](https://github.com/powershell/xcredssp/blob/cafb56015cc5099278b9b86f85272ae665e94f77/Tests/Unit/MSFT_xCredSSP.tests.ps1#L132)):*
> ```PowerShell
>                 }
>                 it 'should have called get' {
>                     Assert-MockCalled -CommandName Get-WSManCredSSP -Times 1
> ```

Why was this test removed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcredssp/14#-:-KeuURePJJH2xcKVloJX:b-fzz2dc)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
What exactly do you mean with end-to-end tests?

I have tested this manually.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[README.md, line 52 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZM58gi4I1rGu-1uo:-KfAtKhdhlkX7_1gO_1C:b-c39owi) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L52)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

How do these changes fix this issue with Test-TargetResource?
</blockquote></details>

I have changed the way the resource is retrieving the delegatecomputers. I am now reading it from the registry instead of using Get-WSManCredSSP.

---

*[Tests/Unit/MSFT_xCredSSP.tests.ps1, line 132 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-Keu_GjBWj59TofmHfg1:-KfAxTtQI9_xS7SDDkJ2:b-bzqygb) ([raw file](https://github.com/powershell/xcredssp/blob/cafb56015cc5099278b9b86f85272ae665e94f77/Tests/Unit/MSFT_xCredSSP.tests.ps1#L132)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

Why was this test removed?
</blockquote></details>

Reviewable doesn't show me the changed info, but I think you are talking about the removal of the call to Get-WSManCredSSP. Since I changed the way the delegatecomputers are retrieved, I no longer use Get-WSManCredSSP, so the test will always fail. That is why I removed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcredssp/14#-:-KfAy5dnTDo89odN-v6B:bruwv0g)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[README.md, line 45 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZWMAvh0Y5tGflqwg:-KfAyNYtRhOZyxwN8Zvx:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L45)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

Can you please also update this section since there is a new parameter?
</blockquote></details>

Done.

---

*[README.md, line 52 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZM58gi4I1rGu-1uo:-KfAyRFpWhWcTpYSfTvN:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L52)):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary><blockquote>

I have changed the way the resource is retrieving the delegatecomputers. I am now reading it from the registry instead of using Get-WSManCredSSP.
</blockquote></details>

Done.

---

*[README.md, line 53 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZ7akEL8Pvnseh-mG:-KfAyQnEXDsVSFnC0zDR:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L53)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

> is Role=Server

if Role=Server
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 13 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuTJBu9vxJXJWG1xk3:-KfAySEvRBO0F3AyLknS:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L13)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

> $DelegateComputers

This variable/parameter doesn't exist here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 146 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuWYD_sCYFXIg11CLW:-KfAyTA_hr_VXyfk8IJc:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L146)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

> ($DelegateComputers)

This should probably be $PSBoundParameters.ContainsKey('DelegateComputers')
Very minor. Only covers the extra case in which someone somehow managed to pass in $null or some mysterious value that translates to false for DelegateComputers
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 311 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuXJ7Uh8WAiNEc_4h0:-KfAyU84Uxb4bNsdCBcN:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L311)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

> Get-ItemProperty -Path $Path -ErrorAction SilentlyContinue

You should probably check that this is not null first here so that it doesn't throw an error that you are trying to call a method/property on a null-valued expression
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.schema.mof, line 7 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuU7JezuMAjYA0_RYi:-KfAyUdR1AysQaWf2YOo:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.schema.mof#L7)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

> is

if
</blockquote></details>

Done.

---

*[Tests/Unit/MSFT_xCredSSP.tests.ps1, line 107 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuXef_1-Q6z-s1vYxo:-KfAyVF6MvM29uFf_nEO:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/Tests/Unit/MSFT_xCredSSP.tests.ps1#L107)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

I know this is how the previous tests here are written, but they are actually wrong as well. These calls to Assert-MockCalled aren't actually checking anything because the call to Set-TargetResource is inside braces ({}) which is considered a different scope. Weird Pester manipulation. Anyways, it's very easy to fix. Just add -Scope 'Context' to the calls to Assert-MockCalled. This will signal to Pester to check for the mock call inside any scope that was called within the Context rather than just the top-level scope of the Context. Like this:

`Assert-MockCalled -CommandName Enable-WSManCredSSP -Times 0 -Scope 'Context'`

The tests are passing at the moment because these are 'not' mock called checks, and since Pester thinks that the function was never even called within the current scope, the mocks were not called either.
</blockquote></details>

Done.

---

*[Tests/Unit/MSFT_xCredSSP.tests.ps1, line 132 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-Keu_GjBWj59TofmHfg1:-KfAyWTHJ7JundYdSCeI:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/cafb56015cc5099278b9b86f85272ae665e94f77/Tests/Unit/MSFT_xCredSSP.tests.ps1#L132)):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary><blockquote>

Reviewable doesn't show me the changed info, but I think you are talking about the removal of the call to Get-WSManCredSSP. Since I changed the way the delegatecomputers are retrieved, I no longer use Get-WSManCredSSP, so the test will always fail. That is why I removed it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcredssp/14)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
By end-to-end tests I mean automated tests that run through the DSC engine like the ones [here](https://github.com/PowerShell/PSDscResources/blob/dev/Tests/Integration/MSFT_Archive.EndToEnd.Tests.ps1). You don't need to worry about it. Manual test pass is fine for this module.

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 52 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuZM58gi4I1rGu-1uo:-KfDzoSIT7kuEPEv2CUc:b-y34a02) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/README.md#L52)):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Ah. Because Test-TargetResource calls Get-TargetResource. Got it.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 311 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuXJ7Uh8WAiNEc_4h0:-KfDyH7DwpEC6WgHsOls:b-zc0tyo) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L311)):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

You need a null check on the result of the function call like this:
```powershell
$itemProperties = Get-ItemProperty -Path $Path -ErrorAction SilentlyContinue
return ($null -ne $itemProperties -and $null -ne $itemProperties.$Name)
```

This is needed because you are ignoring any errors, so if there is an error on the Get-ItemProperty call then the function will return $null.
Then when you try to call $itemProperties.$Name, PowerShell will throw an error that you cannot call a property/method on a null valued expression.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcredssp/14#-:-KfE-ACK9uW-xMmrR5_U:b-ukgh3b)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1, line 311 at r3](https://reviewable.io:443/reviews/powershell/xcredssp/14#-KeuXJ7Uh8WAiNEc_4h0:-Kfwn33FbS2b2RvyjAxG:b-896fix) ([raw file](https://github.com/powershell/xcredssp/blob/1397bdde9940ff67b0a352eed3d7b8b5023ae816/DSCResources/MSFT_xCredSSP/MSFT_xCredSSP.psm1#L311)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

You need a null check on the result of the function call like this:
```powershell
$itemProperties = Get-ItemProperty -Path $Path -ErrorAction SilentlyContinue
return ($null -ne $itemProperties -and $null -ne $itemProperties.$Name)
```

This is needed because you are ignoring any errors, so if there is an error on the Get-ItemProperty call then the function will return $null.
Then when you try to call $itemProperties.$Name, PowerShell will throw an error that you cannot call a property/method on a null valued expression.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcredssp/14)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcredssp/14#-:-Kg1_ZLcRThyJdt1INde:bnfp4nl)*
<!-- Sent from Reviewable.io -->
